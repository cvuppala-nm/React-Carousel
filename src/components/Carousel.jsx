import React, { useRef, useState, useCallback } from 'react';
import '../styles/carousel.scss';
import { MAX_SLIDES, SLIDE_WIDTH } from '../constants/appConstants';
import useResizeListener from '../hooks/useResizeListener';
import useMirrorNodes from '../hooks/useMirrorNodes';

const Carousel = ({ cards, slidesToShow, onAddCard, onCardDelete, onCardUpdate, onUpdateSlidesToShow }) => {
  // slides parent component
  const slidesContainerRef = useRef(null);

  const [slideIndex, setSlideIndex] = useState(2); // 2 - to handle offset for mirror nodes
  const [nodesCount, setNodesCount] = useState(cards.length);

  // Utility function to animate slide movement
  const animateSlide = useCallback((startIndex, endIndex, cb) => {
    const startTime = Date.now();
    const duration = 500; // Duration of the animation in milliseconds
    const startValue = startIndex * SLIDE_WIDTH * -1;
    const endValue = endIndex * SLIDE_WIDTH * -1;
    const changeInValue = endValue - startValue;

    const animate = () => {
      const currentTime = Date.now() - startTime;
      const timeFraction = Math.min(currentTime / duration, 1);
      const easing = timeFraction < 0.5 ? 4 * timeFraction ** 3 : 1 - Math.pow(-2 * timeFraction + 2, 3) / 2; // Ease in-out cubic
      const currentValue = startValue + changeInValue * easing;

      if (slidesContainerRef.current) {
        slidesContainerRef.current.style.transform = `translate3d(${currentValue}px, 0, 0)`;
      }

      if (timeFraction < 1) {
        requestAnimationFrame(animate);
      } else {
        // optional callback when animation is completed
        cb?.();
      }
    };

    requestAnimationFrame(animate);
  }, []);

  // Update nodes count based on the actual children in the DOM
  const updateNodesCount = useCallback(() => {
    if (slidesContainerRef.current) {
      setNodesCount(slidesContainerRef.current.children.length);
    } else {
      setNodesCount(cards.length);
    }
  }, [cards.length]);

  // Custom hook to manage mirror nodes
  useMirrorNodes(slidesContainerRef, cards, updateNodesCount); 

  // Handle window resize with a custom hook to handle responsive view
  useResizeListener(() => {
    const n = Math.floor(window.innerWidth / (SLIDE_WIDTH + 100)); // 100 - some random offset
    const slides = Math.min(MAX_SLIDES, n);
    onUpdateSlidesToShow(slides);
  });

  const listStyles = {
    width: `calc(${Math.min(nodesCount, slidesToShow) * SLIDE_WIDTH}px - 2px)`,
  };

  const trackStyles = {
    transform: `translate3d(${(SLIDE_WIDTH * slideIndex) * -1}px, 0, 0)`,
    width: `calc(${nodesCount * SLIDE_WIDTH}px)`,
  };

  // callback for moving forward to next slide
  const handleNextClick = useCallback(() => {
    // new index for slides
    const newIndex = (slideIndex + 1) % cards.length;
    // request forward animation for navigating beteen slides
    animateSlide(slideIndex, newIndex === 0 ? cards.length : newIndex);
    // set current slide index
    setSlideIndex(newIndex);
  }, [animateSlide, cards.length, slideIndex]);

  // callback for moving back to previous slide
  const handlePrevClick = useCallback(() => {
    // new index for slides
    const newIndex = slideIndex === 0 ? cards.length - 1 : slideIndex - 1;
    // request backward animation for navigating beteen slides
    animateSlide(slideIndex, newIndex);
    // set current slide index
    setSlideIndex(newIndex === 0 ? cards.length : newIndex); // for infinite loop, making index --> cards.length, when newIndex is 0
  }, [animateSlide, cards.length, slideIndex]);

  /** handle on add button click */
  const handleOnAddCard = () => {
    const i = cards?.length
      ? cards[cards.length - 1].index
      : -1;
    
    onAddCard(i + 1);
  };

  /** handle on card update */
  const handleOnCardUpdate = (e, index) => {
    const el = slidesContainerRef.current;

    if (index === 0) {
      el.children[el.children.length - 2].querySelector('h3').textContent = e.target.value;
    }

    if (index === (cards.length - 1)) {
      el.children[1].querySelector('h3').textContent = e.target.value;
    }
    
    onCardUpdate(e, index);
  };

  /** handle on card delete */
  const handleOnCardDelete = (index) => {
    onCardDelete(index);
  };

  return (
    <>
      <div className="carousel-wrapper" role="toolbar">
        <button type="button" className="carousel-nav-btn" aria-label="Previous" onClick={handlePrevClick} disabled={cards.length < slidesToShow}>
          <i className="arrow left"></i>
        </button>
        <div className="carousel-list" style={listStyles}>
          <div className="carousel-track" role="listbox" ref={slidesContainerRef} style={trackStyles}>
            {cards.map((card) => (
              <div key={card.id} className="slide" data-index={card.index} tabIndex="-1">
                <h3>{card.text}</h3>
              </div>
            ))}
          </div>
          <input type="hidden" data-testid="slideIndex" value={slideIndex} />
        </div>
        <button type="button" className="carousel-nav-btn" aria-label="Next" onClick={handleNextClick} disabled={cards.length < slidesToShow}>
          <i className="arrow right"></i>
        </button>
      </div>
      <hr />
      <div className="slides-heading">
        <h3>Slides - ({cards.length})</h3>
        <button type="button" onClick={handleOnAddCard}>Add Slide</button>
      </div>
      <div className="slides-inputs">
        {cards.map((card, index) => (
          <div key={card.id}>
            <input type="text" value={card.text} onChange={(e) => handleOnCardUpdate(e, index)} />
            {cards.length > 1 && (
              <button type="button" onClick={() => handleOnCardDelete(index)}>delete</button>
            )}
          </div>
        ))}
      </div>
    </>
  );
};

export default Carousel;