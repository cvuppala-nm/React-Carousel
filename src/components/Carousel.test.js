import React from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { render, fireEvent, screen } from '@testing-library/react';

import '@testing-library/jest-dom';

import Carousel from './Carousel';
import cardReducer from '../store/cardSlice';

const mockProps = {
  cards: [
    { id: 1, text: '1', index: 0 },
    { id: 2, text: '2', index: 1 },
    { id: 3, text: '3', index: 2 },
  ],
  slidesToShow: 3,
  onAddCard: jest.fn(),
  onCardDelete: jest.fn(),
  onCardUpdate: jest.fn(),
  onUpdateSlidesToShow: jest.fn(),
};

const mockInitialState = {
  card: {
    cards: mockProps.cards,
    slidesToShow: mockProps.slidesToShow,
  },
};

// Utility function to create a mock Redux store
const createMockStore = (initialState) => {
  return configureStore({
    reducer: {
      card: cardReducer,
    },
    preloadedState: initialState,
  });
};

// Updated utility function to render the component with Redux Provider
function renderCarouselWithStore(props, overrideInitialState = {}) {
  const defaultInitialState = {
    card: {
      cards: [
        { id: 1, text: '1', index: 0 },
        { id: 2, text: '2', index: 1 },
        { id: 3, text: '3', index: 2 },
      ],
      slidesToShow: 3,
    },
  };

  const store = createMockStore({ ...defaultInitialState, ...overrideInitialState });

  return render(
    <Provider store={store}>
      <Carousel {...props} />
    </Provider>
  );
}

test('snapshot test for Carousel.tsx', () => {
  const store = createMockStore({ ...mockInitialState });

  const { asFragment } = renderCarouselWithStore(
    mockProps,
    mockInitialState
  );

  expect(asFragment()).toMatchSnapshot();
});

test('renders the carousel with cards from redux store and append mirror nodes', async () => {
  renderCarouselWithStore(mockProps, mockInitialState);
  
  expect(screen.getAllByText('1')).toHaveLength(2); // mirror node
  expect(screen.getAllByText('2')).toHaveLength(3); // mirror node
  expect(screen.getAllByText('3')).toHaveLength(2); // mirror node
});

test('navigates to the next and previous slide', () => {
  renderCarouselWithStore(mockProps, mockInitialState);

  const nextButton = screen.getByLabelText('Next');
  const prevButton = screen.getByLabelText('Previous');

  // Move to next slide
  fireEvent.click(nextButton);
  expect(screen.getByTestId('slideIndex')).toHaveValue('0');

  // Move to previous slide, now it should be enabled
  fireEvent.click(prevButton);
  expect(screen.getByTestId('slideIndex')).toHaveValue('2');
});

test('adds a new card', () => {
  renderCarouselWithStore(mockProps, mockInitialState);
  fireEvent.click(screen.getByText('Add Slide'));
  expect(mockProps.onAddCard).toHaveBeenCalledTimes(1);
});

test('updates a card text', () => {
  renderCarouselWithStore(mockProps, mockInitialState);
  const firstCardInput = screen.getAllByRole('textbox')[0];
  fireEvent.change(firstCardInput, { target: { value: 'Updated Card 1' } });
  expect(mockProps.onCardUpdate).toHaveBeenCalledWith(expect.anything(), 0);
});

test('deletes a card', () => {
  renderCarouselWithStore(mockProps, mockInitialState);
  const deleteButtons = screen.getAllByText('delete');
  fireEvent.click(deleteButtons[0]);
  expect(mockProps.onCardDelete).toHaveBeenCalledTimes(1);
});