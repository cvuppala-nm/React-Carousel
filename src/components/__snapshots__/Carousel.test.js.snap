// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for Carousel.tsx 1`] = `
<DocumentFragment>
  <div
    class="carousel-wrapper"
    role="toolbar"
  >
    <button
      aria-label="Previous"
      class="carousel-nav-btn"
      type="button"
    >
      <i
        class="arrow left"
      />
    </button>
    <div
      class="carousel-list"
      style="width: calc(900px - 2px);"
    >
      <div
        class="carousel-track"
        role="listbox"
        style="transform: translate3d(-600px, 0, 0); width: calc(2100px);"
      >
        <div
          class="slide clone-slide"
          data-index="1"
          tabindex="-1"
        >
          <h3>
            2
          </h3>
        </div>
        <div
          class="slide clone-slide"
          data-index="2"
          tabindex="-1"
        >
          <h3>
            3
          </h3>
        </div>
        <div
          class="slide"
          data-index="0"
          tabindex="-1"
        >
          <h3>
            1
          </h3>
        </div>
        <div
          class="slide"
          data-index="1"
          tabindex="-1"
        >
          <h3>
            2
          </h3>
        </div>
        <div
          class="slide"
          data-index="2"
          tabindex="-1"
        >
          <h3>
            3
          </h3>
        </div>
        <div
          class="slide clone-slide"
          data-index="0"
          tabindex="-1"
        >
          <h3>
            1
          </h3>
        </div>
        <div
          class="slide clone-slide"
          data-index="1"
          tabindex="-1"
        >
          <h3>
            2
          </h3>
        </div>
      </div>
      <input
        data-testid="slideIndex"
        type="hidden"
        value="2"
      />
    </div>
    <button
      aria-label="Next"
      class="carousel-nav-btn"
      type="button"
    >
      <i
        class="arrow right"
      />
    </button>
  </div>
  <hr />
  <div
    class="slides-heading"
  >
    <h3>
      Slides - (3)
    </h3>
    <button
      type="button"
    >
      Add Slide
    </button>
  </div>
  <div
    class="slides-inputs"
  >
    <div>
      <input
        type="text"
        value="1"
      />
      <button
        type="button"
      >
        delete
      </button>
    </div>
    <div>
      <input
        type="text"
        value="2"
      />
      <button
        type="button"
      >
        delete
      </button>
    </div>
    <div>
      <input
        type="text"
        value="3"
      />
      <button
        type="button"
      >
        delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
